### PR TITLE
[JFG-1118]  Exclusive access load balancing in multi kernel mode

### PR DESCRIPTION
**Pull request by a-badaev**

### CODE REVIEW

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadProcess.java [hunk 1]
```diff
@@ -107,15 +109,18 @@ public void stop() {
     public WorkloadStatus getStatus() {
         int started = samplesCountStartedFromTerminatedThreads;
         int finished = samplesCountFinishedFromTerminatedThreads;
+        long emptyTrx = emptyTransactionsFromTerminatedThreads;
```
Comment (SokolAndrey) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadProcess.java]: Pls change the name, I didn't catch what does the Trx mean.

File: chassis/core/src/main/java/com/griddynamics/jagger/invoker/SimpleScenario.java [hunk 2]
```diff
@@ -47,8 +47,9 @@ private SimpleScenario(Invoker<Q, R, E> invoker, Q query, E endpoint, SystemCloc
     }
 
     @Override
-    public void doTransaction() {
+    public boolean doTransaction() {
         invoker().invoke(query, endpoint);
+        return true;
```
Comment (SokolAndrey) [inline chassis/core/src/main/java/com/griddynamics/jagger/invoker/SimpleScenario.java]: Could you provide some explanations?

File: chassis/core/src/main/java/com/griddynamics/jagger/invoker/QueryPoolScenario.java [hunk 3]
```diff
@@ -45,15 +45,19 @@ public QueryPoolScenario(Invoker<Q, R, E> invoker, Iterator<Pair<Q, E>> chunks,
         this.chunks = chunks;
         this.systemClock = systemClock;
     }
-
+    
     @Override
-    public void doTransaction() {
+    public boolean doTransaction() {
         Pair<Q, E> chunk = chunks.next();
+        if (chunk == null) {
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/invoker/QueryPoolScenario.java]: please add comment in which case chunk can be null

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/UserTerminationStrategy.java [hunk 4]
```diff
@@ -86,6 +87,19 @@ public boolean isTerminationRequired(WorkloadExecutionStatus status) {
...
+    
+        boolean everyNodeHasEmptyTransactions = true;
+        for (NodeId nodeId : status.getNodes()) {
+            if (status.getEmptyTransactions(nodeId) == 0) {
+                everyNodeHasEmptyTransactions = false;
+                break;
+            }
+        }
+        if (everyNodeHasEmptyTransactions) {
+            log.info("Request to terminate work. Every node has empty transactions which means there are nothing to do any more");
```
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/UserTerminationStrategy.java]: we are not supposed to know the exact reason of empty transaction here. it is kinda implementation specific

File: chassis/core/src/main/java/com/griddynamics/jagger/invoker/ExclusiveAccessLoadBalancer.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            }
+            
+            super.init();
+    
+            PairSupplier<Q, E> pairSupplier = getPairSupplier();
+            List<Pair<Q, E>> pairList = new ArrayList<>(pairSupplier.size());
+
+            int index = 0;
+            int step = 1;
+            if (kernelInfo != null) { // then from all the provided pairs pick only those under this node's index.
```
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/invoker/ExclusiveAccessLoadBalancer.java]: we are not supposed to know the exact reason  here. it is just a nullable field and we do not constrain it not to be null in that class. so, we have to do the null check

File: chassis/core/src/main/java/com/griddynamics/jagger/invoker/NonCircularExclusiveAccessLoadBalancer.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package com.griddynamics.jagger.invoker;
+
+import com.griddynamics.jagger.util.Pair;
+
+/**
+ * Subclass of {@link ExclusiveAccessLoadBalancer} that provides each pair of Q and E only once (does not circle an iteration).
+ * As a result it works as a finite load balancer over a predefined sequence of pairs.
```
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/invoker/NonCircularExclusiveAccessLoadBalancer.java]: usually docs are not copy pasted from super class to subclass. I explicitly noted which class it extends in the doc  so it is easily to navigate for more details

File: chassis/core/src/main/java/com/griddynamics/jagger/invoker/QueryPoolScenario.java (hunk 3 above)
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/invoker/QueryPoolScenario.java]: we are not supposed to know the exact reason  here. it just can by contract but exact reason is implementation specific

File: chassis/core/src/main/java/com/griddynamics/jagger/invoker/SimpleScenario.java (hunk 2 above)
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/invoker/SimpleScenario.java]: in this implementation we always return true - meaning always have non empty transactions

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/UserTerminationStrategy.java (hunk 4 above)
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/UserTerminationStrategy.java]: Request to terminate work. All planned invocations are finished

File: chassis/core/src/main/java/com/griddynamics/jagger/invoker/QueryPoolScenario.java (hunk 3 above)
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/invoker/QueryPoolScenario.java]: load balancer logics considers there are no more available combinations

File: chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/loadbalancer/JLoadBalancer.java [hunk 7]
```diff
@@ -47,48 +59,81 @@ public static Builder builder(DefaultLoadBalancer loadBalancer) {
...
+            return this;
+        }
+    
+        /**
+         * Optional: If this flag is true the builder will produce a load balancer
+         * which provides each pair only once (does only one iteration over a sequence of those pairs)
+         * @return {@link Builder} this
+         * @see NonCircularExclusiveAccessLoadBalancer
+         */
+        public Builder withNonCirclularIteration() {
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/user/test/configurations/loadbalancer/JLoadBalancer.java]: please rename to withUniqueAccess

File: chassis/configuration/configuration/basic/default.environment.properties [hunk 8]
```diff
@@ -406,4 +406,11 @@ chassis.engine.e1.reporting.session.comparison.decision.maker.warning.deviation.
 # F.e. threshold of 0.8 mean that values for current session should differ from baseline not more than +-80% to avoid errors
 chassis.engine.e1.reporting.session.comparison.decision.maker.fatal.deviation.threshold=0.8
 
+### Section for properties which is going to be used during configuration construction
```
Comment (dlatnikov) [inline chassis/configuration/configuration/basic/default.environment.properties]: please, put it higher in this file. before string
# begin: following section is used for docu generation - Default setup for monitoring metrics

Comment (dlatnikov) [inline chassis/configuration/configuration/basic/default.environment.properties]: please, also add to environment.properties in the archetype

**Pull request merged**